### PR TITLE
chain2: merge pull request #3765 to pybind11

### PR DESCRIPTION
**Pull request by qindazhu**
Merge #3765 to branch pybind11, tested on mini_speech, see [egs/mini_librispeech/s5/local/chain2/run_tdnn.sh](https://github.com/qindazhu/kaldi/blob/7438564d4610205d7818afb820aa9472a32816db/egs/mini_librispeech/s5/local/chain2/run_tdnn.sh)

**Result**
%WER 19.71 [ 3970 / 20138, 298 ins, 994 del, 2678 sub ] exp/chaina/tdnn2c_sp/decode_tgsmall_dev_clean_2/wer_11_0.0
%WER 14.50 [ 2920 / 20138, 332 ins, 598 del, 1990 sub ] exp/chaina/tdnn2c_sp/decode_tglarge_dev_clean_2/wer_10_0.0

**Result from @mrsrikanth**
%WER 21.38 [ 4305 / 20138, 449 ins, 740 del, 3116 sub ] exp/chain2/tdnn1a_sp/decode_tgsmall_dev_clean_2//wer_10_0.0
%WER 15.64 [ 3150 / 20138, 395 ins, 584 del, 2171 sub ] exp/chain2/tdnn1a_sp/decode_tglarge_dev_clean_2//wer_11_0.0

### CODE REVIEW

**Comment by csukuangfj** [comment]
||@qindazhu (kaldi pybind)| @mrsrikanth (kaldi)|
|--|--|---|
|tgsmall|19.71 | 21.38|
|tglarge|14.50| 15.64|

It looks like the training using kaldi pybind  outperforms that of kaldi.

**Comment by qindazhu** [comment]
no pybind here, both result are run with kaldi

> @qindazhu (kaldi pybind)	@mrsrikanth (kaldi)
> tgsmall	19.71	21.38
> tglarge	14.50	15.64
> It looks like the training using kaldi pybind outperforms that of kaldi.

**Comment by danpovey** [comment]
THanks, Haowen; I'll look at this to-morrow.

File: egs/wsj/s5/steps/nnet3/chain2/train.sh [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+      fi
+      $cmd $gpu_cmd_opt $dir/log/diagnostic_${name}.$x.log \
+         nnet3-chain-train2 --use-gpu=$use_gpu \
+            --leaky-hmm-coefficient=$leaky_hmm_coefficient \
+            --xent-regularize=$xent_regularize \
+            --out-of-range-regularize=$out_of_range_regularize \
+            $l2_regularize_opt \
+            --print-interval=10  \
+           "nnet3-copy --learning-rate=$lrate $dir/${x}.raw - |" $den_fst_dir \
+           "ark:nnet3-chain-copy-egs $egs_opts scp:$egs_dir/${name}_subset.scp ark:- | nnet3-chain-merge-egs $multilingual_eg_opts --minibatch-size=1:64 ark:- ark:-|" \
```
Comment (kkm000) [inline egs/wsj/s5/steps/nnet3/chain2/train.sh:191]: `--minibatch-size=1:64` was previously `--minibatch-size=$groups_per_minibatch`.  Was  hardcoding the value intentional?

**Comment by danpovey** [comment]
Hm.  Should probably be called minibatch_size and made an option.

**Comment by kkm000** [comment]
I noticed the source PR has the same thing. The option is declared but not used. It was at some stage in the PR branch. 

I do not understand the "groups per minibatch" concept, is it something new to chain2? This is like 5K lines of new code, I've not gone through it yet.

**Comment by mrsrikanth** [comment]
The reason it is hardcode is because that is how it is implemented in the original chain recipes. but it makes sense to make it an option. Should I make a new PR?

**Comment by kkm000** [comment]
@mrsrikanth, the answer is up to Dan, I do not understand the chain2 yet. But I would appreciate if you could give me a few hints. Very high-level is ok, like what is the general idea of it?

**Comment by jtrmal** [comment]
BTW, my comments from #4001 also apply here, I think.
y.

**Comment by mrsrikanth** [comment]
@kkm000 chain2 is a re-write of the original chain recipes. it avoids using the python libraries in steps/. But the main modification is with respect to the preparation of egs that makes multilingual training with LF-MMI possible.

@jtrmal I don't know how far chain2 has changed after it was merged. So, I'll wait for @danpovey to say if there is anything more left to do for me.

**Comment by jtrmal** [comment]
OK, so you are saying those things do not relate to your changeset? It's
possible, I don't argue, Im just trying to understand.
y.

**Comment by mrsrikanth** [comment]
Those changes (the log format and the cleanup of error file) are definitely from me. I was just trying to figure out that if I had to make a PR to fix those changes, I don't know where to start from (perhaps, kaldi:master?). Or is #4001 trying do something similar?

**Comment by jtrmal** [comment]
I have no idea, honestly...
y.

**Comment by danpovey** [comment]
Regarding groups_per_minibatch: that is a concept that no longer makes
sense, it relates to a feature that was removed from
chain2.  It should just be minibatch_size now.

The minibatch size should be configurable for training; for diagnostics,
though, it can be hard-coded, because memory isn't limiting there.

RE any changes to how the info is printed on the screen: that can be done
later, as far as I am concerned, I think it's not super urgent.

**Pull request merged**